### PR TITLE
[AVR] Fix a crash in AVRInstrInfo::insertIndirectBranch

### LINKED ISSUE #67042: [AVR] crash on attiny85: cannot create long jump without FeatureJMPCALL
The following IR file crashes when compiled with llc-17:

```
$ LLVM_SYMBOLIZER_PATH=llvm-symbolizer-17 llc-17 crash.ll
LLVM ERROR: cannot create long jump without FeatureJMPCALL
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: llc-17 crash.ll
1.      Running pass 'Function Pass Manager' on module 'crash.ll'.
2.      Running pass 'Branch relaxation pass' on function '@__d_vfprintf'
 #0 0x0000fffef6807f98 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xcc7f98)
 #1 0x0000fffef6806260 llvm::sys::RunSignalHandlers() (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xcc6260)
 #2 0x0000fffef680867c (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xcc867c)
 #3 0x0000fffefcbe87d0 (linux-vdso.so.1+0x7d0)
 #4 0x0000fffef57009d0 (/lib/aarch64-linux-gnu/libc.so.6+0x809d0)
 #5 0x0000fffef56ba76c raise (/lib/aarch64-linux-gnu/libc.so.6+0x3a76c)
 #6 0x0000fffef56a74bc abort (/lib/aarch64-linux-gnu/libc.so.6+0x274bc)
 #7 0x0000fffef676a5e8 llvm::report_fatal_error(llvm::Twine const&, bool) (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xc2a5e8)
 #8 0x0000fffef676a434 llvm::report_fatal_error(llvm::Twine const&, bool) (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xc2a434)
 #9 0x0000fffef8868e70 (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0x2d28e70)
#10 0x0000fffef6a36f28 (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xef6f28)
#11 0x0000fffef6b88774 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0x1048774)
#12 0x0000fffef6944bd8 llvm::FPPassManager::runOnFunction(llvm::Function&) (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xe04bd8)
#13 0x0000fffef694a2b4 llvm::FPPassManager::runOnModule(llvm::Module&) (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xe0a2b4)
#14 0x0000fffef6945150 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/usr/lib/llvm-17/bin/../lib/libLLVM-17.so.1+0xe05150)
#15 0x0000aaaad4c30e64 main (/usr/lib/llvm-17/bin/llc+0x10e64)
#16 0x0000fffef56a7780 (/lib/aarch64-linux-gnu/libc.so.6+0x27780)
#17 0x0000fffef56a7858 __libc_start_main (/lib/aarch64-linux-gnu/libc.so.6+0x27858)
#18 0x0000aaaad4c2c5f0 _start (/usr/lib/llvm-17/bin/llc+0xc5f0)
fish: Job 1, 'LLVM_SYMBOLIZER_PATH=llvm-symbo…' terminated by signal SIGABRT (Abort)
```

[crash.zip](https://github.com/llvm/llvm-project/files/12689429/crash.zip)

This was introduced in https://reviews.llvm.org/D155570 (eb33db4f9151b3cb87a45d2135b9decc133f3d09).

CC @jacquesguan @benshi001

### PR DESCRIPTION
**Pull request by benshi001**
Fixes https://github.com/llvm/llvm-project/issues/67042

### CODE REVIEW

**Comment by aykevl** [comment]
To make this test easier to read (and less likely to not work anymore in the future due to unrelated changes), I think it would be better to write the test like branch-relaxation.ll and jmp-long.ll (lots of "nop" instructions).

File: llvm/lib/Target/AVR/AVRInstrInfo.cpp [hunk 1]
```diff
@@ -573,7 +573,7 @@ void AVRInstrInfo::insertIndirectBranch(MachineBasicBlock &MBB,
   if (STI.hasJMPCALL())
     BuildMI(&MBB, DL, get(AVR::JMPk)).addMBB(&NewDestBB);
   else
-    report_fatal_error("cannot create long jump without FeatureJMPCALL");
+    BuildMI(&MBB, DL, get(AVR::RJMPk)).addMBB(&NewDestBB);
```
Comment (jacquesguan) [inline llvm/lib/Target/AVR/AVRInstrInfo.cpp:576]: Should we check the BrOffset? Because relative jump can only handle the range between -2k to+2k.

**Comment by benshi001** [comment]
> To make this test easier to read (and less likely to not work anymore in the future due to unrelated changes), I think it would be better to write the test like branch-relaxation.ll and jmp-long.ll (lots of "nop" instructions between from the jump instruction to the branch target).

Done. I have created a new test `long_jmp.ll` for that, as you suggested.

File: llvm/lib/Target/AVR/AVRInstrInfo.cpp (hunk 1 above)
Comment (benshi001) [reply llvm/lib/Target/AVR/AVRInstrInfo.cpp:576]: No need to do so. The offset is sure to exceed, and incorrect is generated, and the report of "out of range" will be postponed to the assembler or the linker.

**Comment by benshi001** [comment]
This patch just fixes the crash, but incorrect code is still generated, and the assembler or the linker will report "out of range".

A full solution would be introducing "code model", like riscv did 
https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/RISCV/jumptable.ll#L4

when `--model-mode=medium` is specified, we use indirect branch `ijmp` instead of short direct `rjmp`, for AVR devices without long direct jump.
and use indirect jump instead of direct jump in cmodel==medium.

**Review by jacquesguan — APPROVED** [review]
LGTM

**Comment by aykevl** [comment]
> A full solution would be introducing "code model", like riscv did https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/RISCV/jumptable.ll#L4
> 
> when `--model-mode=medium` is specified, we use indirect branch `ijmp` instead of short direct `rjmp`, for AVR devices without long direct jump.

Probably unnecessary. I think the only chips that are limited to `rjmp` are the chips that don't have enough flash to address more memory anyway. For example, the attiny85 has only 8kB flash so the entire flash area can be addressed using `rjmp`. From the ISA documentation:

> Relative jump to an address within PC - 2K +1 and PC + 2K (words). For AVR microcontrollers with pogram memory not exceeding 4K words (8 KB), this instruction can address the entire memory from every address location. See also JMP.

As for the current fix:

> This patch just fixes the crash, but incorrect code is still generated, and the assembler or the linker will report "out of range".

That seems like a reasonable solution to me. I hit this while compiling parts of the libc that will be too large anyway for these small chips, but it's nice to be able to compile those parts anyway. If those large functions are linked, they will result in a linker error anyway because they don't fit the flash space.

File: llvm/test/CodeGen/AVR/branch-relaxation-long.ll [hunk 2]
```diff
@@ -7,6 +8,15 @@
...
 ; CHECK: nop
 ; CHECK: [[BB2]]:
+
+; AVR2-LABEL: relax_to_jmp:
+; AVR2: cpi     r{{[0-9]+}}, 0
+; AVR2: brne    [[BB1:.LBB[0-9]+_[0-9]+]]
+; AVR2: rjmp    [[BB2:.LBB[0-9]+_[0-9]+]]
+; AVR2: [[BB1]]:
+; AVR2: nop
+; AVR2: [[BB2]]:
```
Comment (aykevl) [inline llvm/test/CodeGen/AVR/branch-relaxation-long.ll:18]: I think a comment explaining why `rjmp` is intended would be a good idea, to help future people working on this code.

**Review by aykevl — COMMENTED** [review]
This seems like a reasonable solution. The only thing I'm worried about is what the linker will actually do. Will it report an out of range error? Do we have tests for this?

**Comment by benshi001** [comment]
> This seems like a reasonable solution. The only thing I'm worried about is what the linker will actually do. Will it report an out of range error? Do we have tests for this?

With the attached b.s, an linker error is reported as
```
benshi@BENNSHI-MB0 /tmp % /Applications/Arduino.app/Contents/Java/hardware/tools/avr/bin/avr-gcc b.s -O3 -Wall -mmcu=attiny85
/Applications/Arduino.app/Contents/Java/hardware/tools/avr/bin/../lib/gcc/avr/7.3.0/../../../../avr/bin/ld: a.out section `.text' will not fit in region `text'
/Applications/Arduino.app/Contents/Java/hardware/tools/avr/bin/../lib/gcc/avr/7.3.0/../../../../avr/bin/ld: region `text' overflowed by 614 bytes
collect2: error: ld returned 1 exit status
```

File: llvm/test/CodeGen/AVR/branch-relaxation-long.ll (hunk 2 above)
Comment (benshi001) [reply llvm/test/CodeGen/AVR/branch-relaxation-long.ll:18]: Done. More comments is added.

**Comment by benshi001** [comment]
The b.s is attached here.
[b.s.zip](https://github.com/llvm/llvm-project/files/12738432/b.s.zip)

**Comment by aykevl** [comment]
> With the attached b.s, an linker error is reported as

And what about the LLVM linker (lld)? Does it report a similar error?
(I would assume so, because LLD normally generates such "out of space" errors).

> but it is not easy to add such a test into my patch.

No I don't think it should be part of the test, just something to verify before committing.

In any case, I think this patch is ready to go from my POV, but perhaps others will like to take a look as well.

**Comment by aykevl** [comment]
Checked avr-gcc, and it behaves the same as this patch: it emits an `rjmp` if `jmp` isn't available.
Source: https://godbolt.org/z/dao8sq9hd

File: llvm/test/CodeGen/AVR/branch-relaxation-long.ll [hunk 3]
```diff
@@ -7,6 +8,18 @@
 ; CHECK: [[BB1]]:
 ; CHECK: nop
 ; CHECK: [[BB2]]:
+
+;; A `RJMP` is generated instead of expected `JMP` for AVR2,
+;; and it is up to the linker to report 'out of range' or
+;; 'exceed flash maximum size;.
```
Comment (aykevl) [inline llvm/test/CodeGen/AVR/branch-relaxation-long.ll:14]: Small typo

```suggestion
;; 'exceed flash maximum size'.
```

**Review by aykevl — APPROVED** [review]

File: llvm/test/CodeGen/AVR/branch-relaxation-long.ll (hunk 3 above)
Comment (benshi001) [reply llvm/test/CodeGen/AVR/branch-relaxation-long.ll]: Fixed. Thanks!

**Comment by benshi001** [comment]
> > With the attached b.s, an linker error is reported as
> 
> And what about the LLVM linker (lld)? Does it report a similar error? (I would assume so, because LLD normally generates such "out of space" errors).
> 
> > but it is not easy to add such a test into my patch.
> 
> No I don't think it should be part of the test, just something to verify before committing. If we add a test it would be a separate test to LLD that checks whether an error is generated as we would expect.
> 
> In any case, I think this patch is ready to go from my POV, but perhaps others will like to take a look as well.

LLD checks range for conditional branch: https://github.com/llvm/llvm-project/blob/main/lld/ELF/Arch/AVR.cpp#L233, 

but does not for RJMP: https://github.com/llvm/llvm-project/blob/main/lld/ELF/Arch/AVR.cpp#L240C8-L240C22

I will fix that in another patch.

**Pull request merged**